src/common/reduction/Cluster.cpp

```diff
@@ -196,27 +181,46 @@ double Cluster::time_center2() const {
 double Cluster::coord_utpc(bool weighted) const {
   if (!weighted)
   {
-    return coord_utpc_/cnt_utpc_;
+    return coord_utpc_;
+  }
+
+  double n1 = 0, n2 = 0, n3 = 0, w1 = 0, w2 = 0, w3 = 0;
+  
+  n2 = hits[highest_time_idx].coordinate;
+  w2 = hits[highest_time_idx].weight;
+  if(highest_time_idx > 0) {
+    n1 = hits[highest_time_idx-1].coordinate;
+    w1 = hits[highest_time_idx - 1].weight;
+  } 
+  if(highest_time_idx < static_cast<int>(hits.size() - 1)) {
+    n3 = hits[highest_time_idx+1].coordinate;
+    w3 = hits[highest_time_idx+1].weight;  
   }
-  return coord_mass_utpc_ / weight_utpc_sum_;
+  double pos = (n1*w1*w1 + n2*w2*w2 + n3*w3*w3) / (w1*w1 + w2*w2 + w3*w3);
+  return pos;
 }
 
 double Cluster::time_utpc(bool weighted) const {
   if (!weighted)
   {
-    return time_utpc_/cnt_utpc_;
```

Comment:
Was this a bug? (just curious)

Reply:
In Martin's uTPC for APV algorithm, the latest times were used. If there were several strips with the same latest time, the average of the strip position was calculated. If the option was weighted uTPC, not the average of the strip positions was used, but the weighted average (using the charge). In my new implementation, pure uTPC just takes the strip with the largest time (with ns resolution there are basically never two strips with the same time). The weighted variant takes the strip with the latest time, together with it's one or two neighbours, and calculates center of mass with the charge squared as weight,